File: CS/WinForms-Dashboard-Custom-Properties/Modules/DashboardDescriptionModule.cs
using System;
using System.Windows.Forms;
using DevExpress.DashboardCommon;
using DevExpress.DashboardWin;
using DevExpress.DashboardWin.Localization;
using DevExpress.Utils;
using DevExpress.Utils.Svg;
using DevExpress.XtraBars;
using DevExpress.XtraBars.Ribbon;
using DevExpress.XtraEditors;
using DevExpress.XtraLayout;
using DevExpress.XtraLayout.Utils;

namespace LayoutChange.Modules {
    public class DashboardDescriptionModule {
        public static readonly string PropertyName = "DashboardDescription";
        readonly DashboardDesigner designer;
        readonly SvgImage titleDescriptionImage;

        public DashboardDescriptionModule(DashboardDesigner designer, SvgImage barImage = null, SvgImage titleDescriptionImage = null) {
            this.designer = designer;
            this.titleDescriptionImage = titleDescriptionImage;
            RibbonControl ribbon = (RibbonControl)designer.MenuManager;
            RibbonPage page = ribbon.GetDashboardRibbonPage(DashboardBarItemCategory.None, DashboardRibbonPage.Home);
            RibbonPageGroup group = page.Groups.GetGroupByText(DashboardWinLocalizer.GetString(DashboardWinStringId.RibbonPageDashboardCaption));
            BarButtonItem barItem = CreateBarItem("Dashboard Description", barImage);
            group.ItemLinks.Add(barItem);
            barItem.ItemClick += DescriptionBarItem_Click;
            designer.CustomizeDashboardTitle += Designer_CustomizeDashboardTitle;
            designer.DashboardCustomPropertyChanged += Designer_DashboardCustomPropertyChanged;
        }
        BarButtonItem CreateBarItem(string caption, SvgImage barImage) {
            BarButtonItem barItem = new BarButtonItem();
            barItem.Caption = caption;
            barItem.ImageOptions.SvgImage = barImage;
            return barItem;
        }
        void DescriptionBarItem_Click(object sender, EventArgs e) {
            ShowDescriptionForm();
        }
        void ShowDescriptionForm() {
            string currentCaption = designer.Dashboard.CustomProperties.GetValue(PropertyName);
            LayoutControl layoutControl = new LayoutControl();
            layoutControl.Dock = DockStyle.Fill;
            layoutControl.Name = "layoutControl1";
            LayoutControlItem memoEditLayoutItem = layoutControl.Root.AddItem();
            memoEditLayoutItem.TextLocation = Locations.Top;
            memoEditLayoutItem.Text = "Text:";
            MemoEdit memoEdit = new MemoEdit();
            memoEdit.Text = currentCaption;
            memoEditLayoutItem.Control = memoEdit;
            LayoutControlItem buttonLayoutItem = layoutControl.Root.AddItem();
            buttonLayoutItem.SizeConstraintsType = SizeConstraintsType.Custom;
            buttonLayoutItem.TextVisible = false;
            SimpleButton buttonOk = new SimpleButton();
            buttonOk.Text = "OK";
            buttonOk.DialogResult = DialogResult.OK;
            buttonOk.Click += (sender1, e1) => {
                string modifiedCaption = memoEdit.Text;
                if (currentCaption != modifiedCaption) {
                    string status = !string.IsNullOrEmpty(modifiedCaption) ? "Remove" : "Add";
                    CustomPropertyHistoryItem historyItem = new CustomPropertyHistoryItem(designer.Dashboard, PropertyName, modifiedCaption, status + " the dashboard caption");
                    designer.AddToHistory(historyItem);
                }
            };
            buttonLayoutItem.Control = buttonOk;
            layoutControl.Root.AddItem(new EmptySpaceItem(), buttonLayoutItem, InsertType.Left);
            using(XtraForm form = new XtraForm()) {
                form.Text = "Description";
                form.Size = new System.Drawing.Size(250, 250);
                form.StartPosition = FormStartPosition.CenterParent;
                form.IconOptions.ShowIcon = false;
                form.Controls.Add(layoutControl);
                form.ShowDialog();
            }
        }
        void Designer_CustomizeDashboardTitle(object sender, CustomizeDashboardTitleEventArgs e) {
            string text = designer.Dashboard.CustomProperties.GetValue(PropertyName);
            if(!string.IsNullOrEmpty(text)) {
                DashboardToolbarItem showDataItem = new DashboardToolbarItem("Description",
                    new Action<DashboardToolbarItemClickEventArgs>((args) => {
                        MessageBox.Show(text, "Description");
                    }));
                showDataItem.SvgImage = titleDescriptionImage;
                e.Items.Insert(0, showDataItem);
            }
        }
        void Designer_DashboardCustomPropertyChanged(object sender, CustomPropertyChangedEventArgs e) {
            if(e.Name == PropertyName)
                designer.UpdateDashboardTitle();
        }
    }
}

File: VB/WinForms-Dashboard-Custom-Properties/Modules/DashboardDescriptionModule.vb
Imports System
Imports System.Windows.Forms
Imports DevExpress.DashboardCommon
Imports DevExpress.DashboardWin
Imports DevExpress.DashboardWin.Localization
Imports DevExpress.Utils
Imports DevExpress.Utils.Svg
Imports DevExpress.XtraBars
Imports DevExpress.XtraBars.Ribbon
Imports DevExpress.XtraEditors
Imports DevExpress.XtraLayout
Imports DevExpress.XtraLayout.Utils

Namespace LayoutChange.Modules

    Public Class DashboardDescriptionModule

        Public Shared ReadOnly PropertyName As String = "DashboardDescription"

        Private ReadOnly designer As DashboardDesigner

        Private ReadOnly titleDescriptionImage As SvgImage

        Public Sub New(ByVal designer As DashboardDesigner, ByVal Optional barImage As SvgImage = Nothing, ByVal Optional titleDescriptionImage As SvgImage = Nothing)
            Me.designer = designer
            Me.titleDescriptionImage = titleDescriptionImage
            Dim ribbon As RibbonControl = CType(designer.MenuManager, RibbonControl)
            Dim page As RibbonPage = ribbon.GetDashboardRibbonPage(DashboardBarItemCategory.None, DashboardRibbonPage.Home)
            Dim group As RibbonPageGroup = page.Groups.GetGroupByText(DashboardWinLocalizer.GetString(DashboardWinStringId.RibbonPageDashboardCaption))
            Dim barItem As BarButtonItem = Me.CreateBarItem("Dashboard Description", barImage)
            group.ItemLinks.Add(barItem)
            barItem.ItemClick += AddressOf DescriptionBarItem_Click
            designer.CustomizeDashboardTitle += AddressOf Designer_CustomizeDashboardTitle
            designer.DashboardCustomPropertyChanged += AddressOf Designer_DashboardCustomPropertyChanged
        End Sub

        Private Function CreateBarItem(ByVal caption As String, ByVal barImage As SvgImage) As BarButtonItem
            Dim barItem As BarButtonItem = New BarButtonItem()
            barItem.Caption = caption
            barItem.ImageOptions.SvgImage = barImage
            Return barItem
        End Function

        Private Sub DescriptionBarItem_Click(ByVal sender As Object, ByVal e As EventArgs)
            ShowDescriptionForm()
        End Sub

        Private Sub ShowDescriptionForm()
            Dim currentCaption As String = designer.Dashboard.CustomProperties.GetValue(PropertyName)
            Dim layoutControl As LayoutControl = New LayoutControl()
            layoutControl.Dock = DockStyle.Fill
            layoutControl.Name = "layoutControl1"
            Dim memoEditLayoutItem As LayoutControlItem = layoutControl.Root.AddItem()
            memoEditLayoutItem.TextLocation = Locations.Top
            memoEditLayoutItem.Text = "Text:"
            Dim memoEdit As MemoEdit = New MemoEdit()
            memoEdit.Text = currentCaption
            memoEditLayoutItem.Control = memoEdit
            Dim buttonLayoutItem As LayoutControlItem = layoutControl.Root.AddItem()
            buttonLayoutItem.SizeConstraintsType = SizeConstraintsType.Custom
            buttonLayoutItem.TextVisible = False
            Dim buttonOk As SimpleButton = New SimpleButton()
            buttonOk.Text = "OK"
            buttonOk.DialogResult = DialogResult.OK
            buttonOk.Click += Function(sender1, e1)
                Dim modifiedCaption As String = memoEdit.Text
                If Not Equals(currentCaption, modifiedCaption) Then
                    Dim status As String = If(Not String.IsNullOrEmpty(modifiedCaption), "Remove", "Add")
                    Dim historyItem As CustomPropertyHistoryItem = New CustomPropertyHistoryItem(designer.Dashboard, PropertyName, modifiedCaption, status & " the dashboard caption")
                    designer.AddToHistory(historyItem)
                End If
            End Function
            buttonLayoutItem.Control = buttonOk
            layoutControl.Root.AddItem(New EmptySpaceItem(), buttonLayoutItem, InsertType.Left)
            Using form As XtraForm = New XtraForm()
                form.Text = "Description"
                form.Size = New System.Drawing.Size(250, 250)
                form.StartPosition = FormStartPosition.CenterParent
                form.IconOptions.ShowIcon = False
                form.Controls.Add(layoutControl)
                form.ShowDialog()
            End Using
        End Sub

        Private Sub Designer_CustomizeDashboardTitle(ByVal sender As Object, ByVal e As CustomizeDashboardTitleEventArgs)
            Dim text As String = designer.Dashboard.CustomProperties.GetValue(PropertyName)
            If Not String.IsNullOrEmpty(text) Then
                Dim showDataItem As DashboardToolbarItem = New DashboardToolbarItem("Description", New Action(Of DashboardToolbarItemClickEventArgs)(Sub(args) MessageBox.Show(text, "Description")))
                showDataItem.SvgImage = titleDescriptionImage
                e.Items.Insert(0, showDataItem)
            End If
        End Sub

        Private Sub Designer_DashboardCustomPropertyChanged(ByVal sender As Object, ByVal e As CustomPropertyChangedEventArgs)
            If e.Name Is PropertyName Then designer.UpdateDashboardTitle()
        End Sub
    End Class
End Namespace
